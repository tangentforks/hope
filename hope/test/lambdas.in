(lambda f => lambda x => lambda y => f(x, y)) (+);

(lambda f => lambda x => lambda y => f(x, y)) (+) 3;

(lambda x => lambda y => x) (lambda z => 0);

(lambda x => lambda y => (+)x) (1, 2);

infix COMMA: 3;
data new_type == num COMMA num;

dec left: new_type -> num;
--- left(a COMMA b) <= b;

(lambda x => lambda y => left x)(1 COMMA 2);

infix op: 4;

lambda (op) => lambda x => x op x;

(lambda (op) => lambda x => x op x) (*);
