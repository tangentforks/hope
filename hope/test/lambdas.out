>> lambda x => lambda y => x + y : num -> num -> num
>> lambda y => 3 + y : num -> num
>> lambda y => lambda z => 0 : alpha -> beta -> num
>> lambda y => 1 + 2 : alpha -> num
>> lambda y => left (1 COMMA 2) : alpha -> num
>> lambda (op) => lambda x => x op x : (alpha # alpha -> beta) -> alpha -> beta
>> lambda x => x * x : num -> num
